File: clients/flutter/orisun_flutter/ios/orisun_flutter/Sources/orisun_flutter/OrisunFlutterPlugin.m
#import "OrisunFlutterPlugin.h"

#import <OrisunFlutterMobile/OrisunFlutterMobile.h>

@interface OrisunFlutterPlugin ()
@property(nonatomic, strong) OrisunFluttermobileBridge *bridge;
@end

@implementation OrisunFlutterPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  NSObject<FlutterBinaryMessenger> *messenger = registrar.messenger;
  FlutterMethodChannel *channel = [[FlutterMethodChannel alloc]
      initWithName:@"com.orisunlabs.orisun_flutter/methods"
      binaryMessenger:messenger
      codec:[FlutterStandardMethodCodec sharedInstance]
      taskQueue:[messenger makeBackgroundTaskQueue]];
  OrisunFlutterPlugin *instance = [[OrisunFlutterPlugin alloc] init];
  instance.bridge = OrisunFluttermobileNewBridge();
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  @try {
    NSDictionary<NSString *, id> *arguments = call.arguments ?: @{};
    OrisunFluttermobileBridge *bridge = self.bridge;
    if ([call.method isEqualToString:@"abiVersion"]) {
      result(@([bridge abiVersion]));
    } else if ([call.method isEqualToString:@"open"]) {
      result([bridge open:[self string:arguments key:@"directory"]
            boundariesJSON:[self string:arguments key:@"boundaries"]]);
    } else if ([call.method isEqualToString:@"close"]) {
      result([bridge close:[self integer:arguments key:@"store"]]);
    } else if ([call.method isEqualToString:@"saveEvents"]) {
      result([bridge saveEvents:[self integer:arguments key:@"store"]
                       boundary:[self string:arguments key:@"boundary"]
                     eventsJSON:[self string:arguments key:@"events"]
           expectedPositionJSON:[self string:arguments key:@"expectedPosition"]
                      queryJSON:[self string:arguments key:@"query"]]);
    } else if ([call.method isEqualToString:@"getEvents"]) {
      result([bridge getEvents:[self integer:arguments key:@"store"]
                      boundary:[self string:arguments key:@"boundary"]
              fromPositionJSON:[self string:arguments key:@"fromPosition"]
                     queryJSON:[self string:arguments key:@"query"]
                         count:(long)[self integer:arguments key:@"count"]
                    descending:[self boolean:arguments key:@"descending"]]);
    } else if ([call.method isEqualToString:@"getLatestByCriteria"]) {
      result([bridge getLatestByCriteria:[self integer:arguments key:@"store"]
                                 boundary:[self string:arguments key:@"boundary"]
                                queryJSON:[self string:arguments key:@"query"]]);
    } else if ([call.method isEqualToString:@"createBoundaryIndex"]) {
      result([bridge createBoundaryIndex:[self integer:arguments key:@"store"]
                                boundary:[self string:arguments key:@"boundary"]
                                    name:[self string:arguments key:@"name"]
                              fieldsJSON:[self string:arguments key:@"fields"]
                          conditionsJSON:[self string:arguments key:@"conditions"]
                              combinator:[self string:arguments key:@"combinator"]]);
    } else if ([call.method isEqualToString:@"dropBoundaryIndex"]) {
      result([bridge dropBoundaryIndex:[self integer:arguments key:@"store"]
                              boundary:[self string:arguments key:@"boundary"]
                                  name:[self string:arguments key:@"name"]]);
    } else if ([call.method isEqualToString:@"subscribe"]) {
      result([bridge subscribe:[self integer:arguments key:@"store"]
                       boundary:[self string:arguments key:@"boundary"]
                 subscriberName:[self string:arguments key:@"subscriberName"]
              afterPositionJSON:[self string:arguments key:@"afterPosition"]
                      queryJSON:[self string:arguments key:@"query"]]);
    } else if ([call.method isEqualToString:@"subscriptionNext"]) {
      int64_t subscription = [self integer:arguments key:@"subscription"];
      int64_t timeoutMillis = [self integer:arguments key:@"timeoutMillis"];
      dispatch_async(dispatch_get_global_queue(QOS_CLASS_USER_INITIATED, 0), ^{
        result([bridge subscriptionNext:subscription timeoutMillis:timeoutMillis]);
      });
    } else if ([call.method isEqualToString:@"stopSubscription"]) {
      result([bridge subscriptionStop:[self integer:arguments key:@"subscription"]]);
    } else {
      result(FlutterMethodNotImplemented);
    }
  } @catch (NSException *exception) {
    result([FlutterError errorWithCode:@"orisun_native_error"
                               message:exception.reason
                               details:exception.callStackSymbols]);
  }
}

- (NSString *)string:(NSDictionary<NSString *, id> *)arguments key:(NSString *)key {
  id value = arguments[key];
  if (![value isKindOfClass:[NSString class]]) {
    @throw [NSException exceptionWithName:NSInvalidArgumentException
                                   reason:[NSString stringWithFormat:@"%@ must be a string", key]
                                 userInfo:nil];
  }
  return value;
}

- (int64_t)integer:(NSDictionary<NSString *, id> *)arguments key:(NSString *)key {
  id value = arguments[key];
  if (![value isKindOfClass:[NSNumber class]]) {
    @throw [NSException exceptionWithName:NSInvalidArgumentException
                                   reason:[NSString stringWithFormat:@"%@ must be an integer", key]
                                 userInfo:nil];
  }
  return [value longLongValue];
}

- (BOOL)boolean:(NSDictionary<NSString *, id> *)arguments key:(NSString *)key {
  id value = arguments[key];
  if (![value isKindOfClass:[NSNumber class]]) {
    @throw [NSException exceptionWithName:NSInvalidArgumentException
                                   reason:[NSString stringWithFormat:@"%@ must be a boolean", key]
                                 userInfo:nil];
  }
  return [value boolValue];
}

- (void)dealloc {
  [self.bridge shutdown];
}

@end

File: clients/flutter/orisun_flutter/android/src/main/kotlin/com/orisunlabs/orisun_flutter/OrisunFlutterPlugin.kt
package com.orisunlabs.orisun_flutter

import com.orisunlabs.orisun.fluttermobile.Bridge
import com.orisunlabs.orisun.fluttermobile.Fluttermobile
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.StandardMethodCodec

class OrisunFlutterPlugin : FlutterPlugin, MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var bridge: Bridge

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        bridge = Fluttermobile.newBridge()
        val taskQueue = binding.binaryMessenger.makeBackgroundTaskQueue(
            BinaryMessenger.TaskQueueOptions().setIsSerial(false),
        )
        channel = MethodChannel(
            binding.binaryMessenger,
            "com.orisunlabs.orisun_flutter/methods",
            StandardMethodCodec.INSTANCE,
            taskQueue,
        )
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        try {
            val args = call.arguments<Map<String, Any?>>() ?: emptyMap()
            val response: Any = when (call.method) {
                "abiVersion" -> bridge.abiVersion()
                "open" -> bridge.open(args.string("directory"), args.string("boundaries"))
                "close" -> bridge.close(args.long("store"))
                "saveEvents" -> bridge.saveEvents(
                    args.long("store"),
                    args.string("boundary"),
                    args.string("events"),
                    args.string("expectedPosition"),
                    args.string("query"),
                )
                "getEvents" -> bridge.getEvents(
                    args.long("store"),
                    args.string("boundary"),
                    args.string("fromPosition"),
                    args.string("query"),
                    args.long("count"),
                    args.boolean("descending"),
                )
                "getLatestByCriteria" -> bridge.getLatestByCriteria(
                    args.long("store"),
                    args.string("boundary"),
                    args.string("query"),
                )
                "createBoundaryIndex" -> bridge.createBoundaryIndex(
                    args.long("store"),
                    args.string("boundary"),
                    args.string("name"),
                    args.string("fields"),
                    args.string("conditions"),
                    args.string("combinator"),
                )
                "dropBoundaryIndex" -> bridge.dropBoundaryIndex(
                    args.long("store"),
                    args.string("boundary"),
                    args.string("name"),
                )
                "subscribe" -> bridge.subscribe(
                    args.long("store"),
                    args.string("boundary"),
                    args.string("subscriberName"),
                    args.string("afterPosition"),
                    args.string("query"),
                )
                "subscriptionNext" -> bridge.subscriptionNext(
                    args.long("subscription"),
                    args.long("timeoutMillis"),
                )
                "stopSubscription" -> bridge.subscriptionStop(args.long("subscription"))
                else -> {
                    result.notImplemented()
                    return
                }
            }
            result.success(response)
        } catch (error: Throwable) {
            result.error("orisun_native_error", error.message, error.stackTraceToString())
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        bridge.shutdown()
    }
}

private fun Map<String, Any?>.string(key: String): String = this[key] as? String
    ?: throw IllegalArgumentException("$key must be a string")

private fun Map<String, Any?>.long(key: String): Long = (this[key] as? Number)?.toLong()
    ?: throw IllegalArgumentException("$key must be an integer")

private fun Map<String, Any?>.boolean(key: String): Boolean = this[key] as? Boolean
    ?: throw IllegalArgumentException("$key must be a boolean")
